lib/rubocop/cop/lint/constant_reassignment.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Checks for constant reassignments.
+      #
+      # Emulates Ruby's runtime warning "already initialized constant X"
+      # when a constant is reassigned in the same file and namespace using the
+      # `NAME = value` syntax.
+      #
+      # The cop cannot catch all offenses, like, for example, when a constant
+      # is reassigned in another file, or when using metaprogramming (`Module#const_set`).
+      #
+      # The cop also tracks constant removal using `Module#remove_const` with a symbol
+      # argument.
+      #
+      # @example
+      #   # bad
+      #   X = :foo
+      #   X = :bar
+      #
+      #   # bad
+      #   class A
+      #     X = :foo
+      #     X = :bar
+      #   end
+      #
+      #   # bad
+      #   module A
+      #     X = :foo
+      #     X = :bar
+      #   end
+      #
+      #   # good - keep only one assignment
+      #   X = :bar
+      #
+      #   class A
+      #     X = :bar
+      #   end
+      #
+      #   module A
+      #     X = :bar
+      #   end
+      #
+      #   # good - use OR assignment
+      #   X = :foo
+      #   X ||= :bar
```

Comment:
Note, if a constant is assigned to `nil`, this would be a false negative:
```ruby
X = nil

X ||= :foo
# warning: already initialized constant X
```